core/libdeno.rs

```diff
@@ -14,11 +24,713 @@ use std::slice;
+    &specifier_str,
+    &referrer_name_str,
+    import_id,
+  );
+
+  &mut *scope.escape(promise)
+}
+
+extern "C" fn host_initialize_import_meta_object_callback(
+  context: v8::Local<v8::Context>,
+  module: v8::Local<v8::Module>,
+  meta: v8::Local<v8::Object>,
+) {
+  let mut cbs = v8::CallbackScope::new(context);
+  let mut hs = v8::HandleScope::new(cbs.enter());
+  let scope = hs.enter();
+  let mut isolate = scope.isolate();
+  let deno_isolate: &mut DenoIsolate =
+    unsafe { &mut *(isolate.get_data(0) as *mut DenoIsolate) };
+
+  let id = module.get_identity_hash();
+  assert_ne!(id, 0);
+
+  let info = deno_isolate.get_module_info(id).expect("Module not found");
+
+  let main = if info.main {
+    v8::Boolean::new(scope, true)
+  } else {
+    v8::Boolean::new(scope, false)
+  };
+
+  meta.create_data_property(
+    context,
+    v8::String::new(scope, "url").unwrap().into(),
+    v8::String::new(scope, &info.name).unwrap().into(),
+  );
+  meta.create_data_property(
+    context,
+    v8::String::new(scope, "main").unwrap().into(),
+    main.into(),
+  );
+}
+
+extern "C" fn message_callback(
+  message: v8::Local<v8::Message>,
+  exception: v8::Local<v8::Value>,
+) {
+  let mut message: v8::Local<v8::Message> =
+    unsafe { std::mem::transmute(message) };
+  let isolate = message.get_isolate();
+  let deno_isolate: &mut DenoIsolate =
+    unsafe { &mut *(isolate.get_data(0) as *mut DenoIsolate) };
+  let mut locker = v8::Locker::new(isolate);
+  let mut hs = v8::HandleScope::new(&mut locker);
+  let scope = hs.enter();
+  assert!(!deno_isolate.context_.is_empty());
+  let mut context = deno_isolate.context_.get(scope).unwrap();
+
+  // TerminateExecution was called
+  if isolate.is_execution_terminating() {
+    let u = v8::new_undefined(scope);
+    deno_isolate.handle_exception(scope, context, u.into());
+    return;
+  }
+
+  let json_str = deno_isolate.encode_message_as_json(scope, context, message);
+  deno_isolate.last_exception_ = Some(json_str);
+}
+
+extern "C" fn promise_reject_callback(msg: v8::PromiseRejectMessage) {
+  #[allow(mutable_transmutes)]
+  let mut msg: v8::PromiseRejectMessage = unsafe { std::mem::transmute(msg) };
+  let mut isolate = msg.isolate();
+  let deno_isolate: &mut DenoIsolate =
+    unsafe { &mut *(isolate.get_data(0) as *mut DenoIsolate) };
+  let mut locker = v8::Locker::new(isolate);
+  assert!(!deno_isolate.context_.is_empty());
+  let mut hs = v8::HandleScope::new(&mut locker);
+  let scope = hs.enter();
+  let mut context = deno_isolate.context_.get(scope).unwrap();
+  context.enter();
+
+  let promise = msg.get_promise();
+  let promise_id = promise.get_identity_hash();
+
+  match msg.get_event() {
+    v8::PromiseRejectEvent::PromiseRejectWithNoHandler => {
+      let error = msg.get_value();
+      let mut error_global = v8::Global::<v8::Value>::new();
+      error_global.set(scope, error);
+      deno_isolate
+        .pending_promise_map_
+        .insert(promise_id, error_global);
```

Comment:
@ry @piscisaureus `msg.get_value()` is an example of empty `Local<>` handle - originally (ie. in `libdeno`) promise was assigned to variable at the top of the function - I tried doing the same here and it failed. To be more precise, if event is `PromiseHandlerAddedAfterReject` then `mgs.get_promise()` returns `Local<Promise>` which is empty, in our case it crashes because of the `unwrap()` in `get_promise`.

Anyway, this works now, just wanted to let you know I encountered `Local<>` that actually has an empty handle

Reply:
> Anyway, this works now, just wanted to let you know I encountered Local<> that actually has an empty handle

I'm aware that this happens. If you encounter one of these cases, feel free to change the API by swapping out the `Local<T>` for an `Option<Local<T>>`.